### ISSUE TITLE
process stucks/hangs and the memory usage increases up to 4.6Gib

### ISSUE THREAD

**Issue by ArtieReus** [github-changelog-generator/github-changelog-generator#651]
Hi again,
i was very happy to know that my pull request #616 got merged a few days ago so I could get rid of my fork. Sadly I figured out that using the last state of the repo the process stucks/hangs and the memory usage increased to 4.6Gib.

![bildschirmfoto 2018-05-08 um 13 35 17](https://user-images.githubusercontent.com/1876079/39758981-b1289f00-52d1-11e8-8932-6739d3f92501.png)

So I decided to use the [git-bisect](https://git-scm.com/docs/git-bisect) to figure out when was the bug introduced. As starting point i put following commit f977cd107243afa382be0f69db27119d0450985f which was added some days before I opened my fork and went the whole way through testing every step. 

    $ git bisect start
    $ git bisect bad
    $ git bisect good f977cd107243afa382be0f69db27119d0450985f

At each step i build the gem and tested agains my [project](https://github.com/sapcc/elektra/tree/master/contrib/elektra-changelog-builder) locally and not from the docker.

As a result i got following:

    372875f74838e5d7ab38bc337ec4333b540c13cd is the first bad commit
    commit 372875f74838e5d7ab38bc337ec4333b540c13cd
    Author: Hunter Haugen <hunter@puppet.com>
    Date:   Thu Feb 8 16:28:39 2018 -0800

    Use git history to find PRs in a tag instead of time

    Fixes #617

    Before this change PRs were associated with tags by the merge date of
    the PR, regardless of git branching history. In this change the oldest
    tag in which the merge commit of each PR occurs in is determined by
    using the Github compare API. That association is then used to sort PRs
    into their respective release.

    This change uses caching of commit history fetching and comparisons for
    speedup.

    :040000 040000 85ea95fe6eee626debc717ad31c2e56b790bd5bf 
    ace182352a34c7ece0d0f4c76e0e5b068d152b99 M	lib
    :040000 040000 7bf68589dad03500e971f992cfa4194824e60177 
    bd10aac7205bb84900a4a07a7ea3c3a634712e5c M	spec

I didn't had time to look into the commit. I will again open a fork rebasing to the commit before the bug and cherry pick the commits from my pull request... until we find a solution.

Best Arturo

**Comment by hunner** [comment]
That's not great. I am able to replicate this with `bundle exec ./bin/github_changelog_generator --user sapcc --project elektra` on the current master (7fe9adb7439ae6468dc487e4273637ac568280bd). `du -sh` says the repo is 29M cloned on-disk.

The repo has 59 closed issues, 199 closed PRs, and 4617 commits.

I `^C`d when the memory usage was ~3G and got this stacktrace:
```
.../lib/github_changelog_generator/generator/generator_fetcher.rb:46:in `join': Interrupt
	from .../lib/github_changelog_generator/generator/generator_fetcher.rb:46:in `each'
	from .../lib/github_changelog_generator/generator/generator_fetcher.rb:46:in `block in detect_actual_closed_dates'
	from .../lib/github_changelog_generator/generator/generator_fetcher.rb:41:in `each'
	from .../lib/github_changelog_generator/generator/generator_fetcher.rb:41:in `each_slice'
	from .../lib/github_changelog_generator/generator/generator_fetcher.rb:41:in `detect_actual_closed_dates'
	from .../lib/github_changelog_generator/generator/generator.rb:150:in `fetch_issues_and_pr'
	from .../lib/github_changelog_generator/generator/generator.rb:47:in `compound_changelog'
	from .../lib/github_changelog_generator.rb:35:in `run'
	from ./bin/github_changelog_generator:5:in `<main>'
```

I removed the threading around `find_closed_dates_by_commit()` and it ran with no problems (though every PR before number 143 has no associated merge commit; perhaps a history rebasing happened around then). So it must be something with the way the closing dates by commit are found when part of threading.

Oh yeah, apparently https://github.com/github-changelog-generator/github-changelog-generator/blob/7fe9adb7439ae6468dc487e4273637ac568280bd/lib/github_changelog_generator/generator/generator_fetcher.rb#L42-L46 is spawning one thread per issue rather than one thread per batch of issues.

**Comment by ArtieReus** [comment]
Thx for the root cause analysis!

**Comment by hunner** [comment]
@ArtieReus If the fix in #653 fixes the issue for you I'd love to know!

**Comment by ArtieReus** [comment]
@hunner let me try the fix against my repo tomorrow and I will post the results. Thx

**Comment by ArtieReus** [comment]
@hunner I test the fix #653 and it works good against the elektra repo. It takes a bit longer than before (before it is meant status from end of january), but it works through:

    Done!
    Generated log placed in ///host/CHANGELOG.md
   
    real	2m21.545s
    user	0m0.245s
    sys	0m0.096s

System usage is around 800Mib which is a bit but ok:
![bildschirmfoto 2018-05-16 um 10 48 29](https://user-images.githubusercontent.com/1876079/40106875-354c45a4-58f7-11e8-8bb1-60ee2e97aa11.png)

Thx
